components/application/application.js

```diff
@@ -0,0 +1,204 @@
+    try{
+        // find application by id and user
+        let a = await Application.findOne({
+            _id         : req.params.id,
+            requestedBy : req.auth.id,
+        })
+        if(!a){
+            return next(
+                new Error("No such application for the current user.")
+            )
+        }
+
+        // we don't let the user see very single detail.
+        let result = {
+            amount        : a.amount,
+            payments      : a.payments,
+            paymentAmount : a.paymentAmount,
+            reason        : a.reason,
+            description   : a.description,
+            status        : a.status,
+            requestedAt   : a.createdAt
+        }
+
+        if("rejected" === a.status){
+            result.rejectedReason = a.rejectedReason
+        }
+
+        return res.status(200).json(result)
+    }catch(e){
+        return next(e)
+    }
+}
+
+/**
+ * Cancels an application. User initiated.
+ * When this happens, we need to give the user a "cooldown period" before he/she can apply again.
+ */
+async function postCancelApplication(req,res,next){
+    try{
+        let application = await Application.findOne({
+            // this is necessary.
+            _id         : req.params.id,
+            requestedBy : req.auth.id,
+            status      : "pending"
+        })
+
+        // did not find it.
+        if(application == null){
+            return next(
+                new Error("No pending application with that id for the current user.")
+            )
+        }
+
+        // change status to cancelled.
+        application.status = "cancelled"
+        await application.save()
+
+        return res.status(200).json({
+            msg : "Application cancelled successfully."
+        })
+
+        // TODO: notify the system there was a cancellation
+    }catch(e){
+        return next(e)
+    }
+}
+
+
+
+
+function adminGetAllApplications(req,res){
+    res.status(200).json({
+        msg : "All applications."
+    })
+}
+
+function adminPostApproveApplication(req,res){
+    res.status(200).json({
+        msg : "Approving application "+req.params.id
+    })
+}
+
+function adminPostRejectApplication(req,res){
+    res.status(200).json({
+        msg    : "Rejecting application "+req.params.id,
+        reason : req.body.reason
+    })
+}
+
+
+export const applicationValidator = [
+    protectedRoute,
+    body('amount').isNumeric(),
+    body('payments').isNumeric(),
+    body('paymentAmount').isNumeric(),
+    body('reason').exists().trim().escape(),
+    body('description').exists().trim().escape(),
+    validationGuard
+]
+
+export default function(app){
+    // user procedures
+    app.post("/application/apply"      , applicationValidator,postMakeApplication)
+    app.post("/application/cancel/:id" , protectedRoute      ,postCancelApplication)
+    app.get ("/application/view/:id"   , protectedRoute      ,getViewApplication)

```

Comment:
We could create validators for these :id parameters as well

Reply:
We could but that would involve obtaining an application inside a middleware and placing it inside the request context. This can get quite confusing ("where the hell is this coming from?") so I'd prefer to limit its usage to authentication